app/services/move_work_package_service.rb

```diff
@@ -0,0 +1,191 @@
+
+    bulk_assign_attributes(modified_work_package, attributes)
+
+    modified_work_package
+  end
+
+  def copy(modified_work_package, attributes, options)
+    set_default_values_on_copy(modified_work_package, attributes)
+
+    return false unless modified_work_package.save
+
+    create_and_save_journal_note modified_work_package, options[:journal_note]
+
+    true
+  end
+
+  def move(modified_work_package, new_project, options)
+    if options[:journal_note]
+      modified_work_package.add_journal user, options[:journal_note]
+    end
+
+    return false unless modified_work_package.save
+
+    move_time_entries(modified_work_package, new_project)
+
+    return false unless move_children(modified_work_package, new_project, options)
+
+    true
+  end
+
+  def move_to_project(work_package, new_project)
+    if new_project &&
+       work_package.project_id != new_project.id &&
+       allowed_to_move_to_project?(new_project)
+
+      work_package.delete_relations(work_package)
+
+      reassign_category(work_package, new_project)
+
+      # Keep the fixed_version if it's still valid in the new_project
+      unless new_project.shared_versions.include?(work_package.fixed_version)
+        work_package.fixed_version = nil
+      end
+
+      work_package.project = new_project
+
+      enforce_cross_project_settings(work_package)
+    end
+  end
+
+  def move_to_type(work_package, new_type)
+    if new_type
+      work_package.type = new_type
+      work_package.reset_custom_values!
+    end
+  end
+
+  def bulk_assign_attributes(work_package, attributes)
+    # Allow bulk setting of attributes on the work_package
+    if attributes
+      # before setting the attributes, we need to remove the move-related fields
+      work_package.attributes =
+        attributes.except(:copy, :new_project_id, :new_type_id, :follow, :ids)
+          .reject { |_key, value| value.blank? }
+    end # FIXME this eliminates the case, where values shall be bulk-assigned to null,
+    # but this needs to work together with the permit
+  end
+
+  def set_default_values_on_copy(work_package, attributes)
+    work_package.author = user
+
+    custom_field_values = self.work_package.custom_field_values.inject({}) { |h, v|
+      h[v.custom_field_id] = v.value
+      h
+    }
+
+    work_package.custom_field_values = custom_field_values
+
+    assign_status_or_default(work_package, attributes[:status_id])
+  end
+
+  def move_children(work_package, new_project, options)
+    work_package.children.each do |child|
+      child_service = self.class.new(child, user)
+      unless child_service.call(new_project, nil, options.merge(no_transaction: true))
+        # Move failed and transaction was rollback'd
+        return false
+      end
+    end
+
+    true
+  end
+
+  def move_time_entries(work_package, new_project)
+    # Manually update project_id on related time entries
+    TimeEntry.update_all("project_id = #{new_project.id}", work_package_id: work_package.id)
```

Comment:
I am wondering if cost entries (from the costs plugin) get updated in the same way.
And if yes: Why and where?